Server Preparation for Ubuntu 22.04 (#369)
owncloud/docs-server: modules/admin_manual/pages/installation/quick_guides/ubuntu_22_04.adoc

echo $sec_admin_pwd > /etc/.sec_admin_pwd.txt

sec_db_pwd=$(openssl rand -base64 18)
echo $sec_db_pwd > /etc/.sec_db_pwd.txt
----

=== Update Your System

First, ensure that all the installed packages are entirely up to date and that PHP is available in the APT repository.
To do so, follow the instructions below:

[source,bash]
----
apt update && \
  apt upgrade -y
----

=== Create the occ Helper Script

Create a helper script to simplify running xref:configuration/server/occ_command.adoc[occ commands]:

[source,bash]
----
FILE="/usr/local/bin/occ"
cat <<EOM >$FILE
#! /bin/bash
cd /var/www/owncloud
sudo -E -u www-data /usr/bin/php /var/www/owncloud/occ "\$@"
EOM
----

Make the helper script executable:

[source,bash]
----
chmod +x $FILE
----

=== Install the Required Packages

[source,bash]
----
sudo add-apt-repository ppa:ondrej/php -y
sudo apt update && sudo apt upgrade

apt install -y \
  apache2 \
  libapache2-mod-php7.4 \
  mariadb-server \
  openssl redis-server wget \
  php7.4 php7.4-imagick php7.4-common \ 
  php7.4-curl php7.4-gd php7.4-imap \
  php7.4-intl php7.4-json php7.4-mbstring \
  php7.4-mysql php7.4-ssh2 php7.4-xml \
  php7.4-zip php7.4-apcu php7.4-redis \

>   php7.4-ldap php-opcache \

Author: It should be php7.4-opcache otherwise you will get an `Package 'php-opcache' has no installation candidate` error

>   php7.4-ldap php-opcache \

Reviewer: ```suggestion
  php7.4-ldap php7.4-opcache php7.4-phpseclib\
```
fix missing 7.4 and harmonizing

>   php7.4-ldap php7.4-opcache php7.4-phpseclib\

Reviewer 2: On Ubuntu 22.04 with Ondrej repos I cannot find php7.4-phpseclib

Should that be `php-phpseclib` instead? That exists, but is probably meant for php8, not sure if that also works with 7.4 ...

Author: @jnweiger You are right, `php7.4-phpseclib` does not exist on Ondrej repo, I may have done an error while documenting. Not sure if we want to document at this point `php-phpseclib` as required package

Reviewer: Note, seems that php-seclib is part of the ondrej package, see: https://github.com/oerdnj/deb.sury.org/issues/445
Needs clarification and raises the question now is how is it named so it can be included properly...?

Reviewer: ok, I think I got the solution 😄 and I will update the PR accordingly

From: https://phpseclib.com/docs/why
```
Why phpseclib?
phpseclib provides pure-PHP implementations of SSH2, SFTP, RSA, DSA, Elliptic Curves, AES, ChaCha20, X.509, CSR, CRL, SPKAC, etc.

Portability
The only requirement that phpseclib 3.0 has is that you must be using PHP 5.6+.

Extensions like bcmath, gmp, libsodium and openssl, if they're available, for speed, but they're not required.
```

And when installing it with `sudo apt install php-phpseclib` you will find the follonwing files:
```
dpkg -L php-phpseclib

/.
/usr
/usr/share
/usr/share/doc
/usr/share/doc/php-phpseclib
/usr/share/doc/php-phpseclib/changelog.Debian.gz
/usr/share/doc/php-phpseclib/copyright
/usr/share/doc/php-phpseclib/examples
/usr/share/doc/php-phpseclib/examples/openssl.cnf
/usr/share/php
/usr/share/php/phpseclib
/usr/share/php/phpseclib/Crypt
/usr/share/php/phpseclib/Crypt/AES.php
/usr/share/php/phpseclib/Crypt/Base.php
/usr/share/php/phpseclib/Crypt/Blowfish.php
/usr/share/php/phpseclib/Crypt/DES.php
/usr/share/php/phpseclib/Crypt/Hash.php
/usr/share/php/phpseclib/Crypt/RC2.php
/usr/share/php/phpseclib/Crypt/RC4.php
/usr/share/php/phpseclib/Crypt/RSA.php
/usr/share/php/phpseclib/Crypt/Random.php
...
```

Reviewer: update applied, all good

----
FILE="/etc/apache2/sites-available/owncloud.conf"
cat <<EOM >$FILE
<VirtualHost *:80>
# uncommment the line below if variable was set
#ServerName $my_domain 
DirectoryIndex index.php index.html
DocumentRoot /var/www/owncloud
<Directory /var/www/owncloud>
  Options +FollowSymlinks -Indexes
  AllowOverride All
  Require all granted

 <IfModule mod_dav.c>
  Dav off
 </IfModule>

 SetEnv HOME /var/www/owncloud
 SetEnv HTTP_HOME /var/www/owncloud
</Directory>
</VirtualHost>
EOM
----

==== Enable the Virtual Host Configuration

[source,bash]
----
a2dissite 000-default
a2ensite owncloud.conf
----

=== Configure the Database

IMPORTANT: It's recommended to execute `mysql_secure_installation` to secure the mariadb installation and set a strong password for the database user.

Ensure transaction-isolation level is set and performance_schema on.

[source,bash]
----
sed -i "/\[mysqld\]/atransaction-isolation = READ-COMMITTED\nperformance_schema = on" /etc/mysql/mariadb.conf.d/50-server.cnf
systemctl start mariadb
mysql -u root -e "CREATE DATABASE IF NOT EXISTS owncloud; \
GRANT ALL PRIVILEGES ON owncloud.* \
  TO owncloud@localhost \
  IDENTIFIED BY '${sec_db_pwd}'";
----

> It is recommended to run mysqltuner script to analyse database configuration after running with load for several days.

Reviewer 3: ```suggestion
It is recommended to run the mysqltuner script to analyze the database configuration after running with load for several days.
```

==== Enable the Recommended Apache Modules

[source,bash]
----
a2enmod dir env headers mime rewrite setenvif
systemctl restart apache2
----

== Installation

=== Download ownCloud

[source,bash,subs="attributes+"]
----
cd /var/www/
wget {oc-complete-base-url}/{oc-complete-name}.tar.bz2 && \
tar -xjf {oc-complete-name}.tar.bz2 && \
chown -R www-data. owncloud
----

=== Install ownCloud

> IMPORTANT: Remember to set a strong password for your owncloud admin user and provide the previously set password for the database user as the `--database-pass` argument.

Reviewer 3: ```suggestion
IMPORTANT: Remember to set a strong password for your ownCloud admin user and provide the previously set password for the database user as the `--database-pass` argument.
```

[source,bash]
----
occ maintenance:install \
    --database "mysql" \
    --database-name "owncloud" \
    --database-user "owncloud" \
    --database-pass ${sec_db_pwd} \
    --data-dir "/var/www/owncloud/data" \
    --admin-user "admin" \
    --admin-pass ${sec_admin_pwd}
----

=== Configure ownCloud's Trusted Domains

[source,bash]
----
my_ip=$(hostname -I|cut -f1 -d ' ')
occ config:system:set trusted_domains 1 --value="$my_ip"
occ config:system:set trusted_domains 2 --value="$my_domain"
----

=== Configure the cron Jobs

Set your background job mode to cron:

[source,bash]
----
occ background:cron
----

> Configure the execution of the cron job to every 15 min and the cleanup of chunks every night at 2 am:

Reviewer 3: ```suggestion
Set the execution of the cron job to every 15 minutes and the cleanup of chunks every night at 2 am:
```

[source,bash]
----
echo "*/15  *  *  *  * /var/www/owncloud/occ system:cron" \
  | sudo -u www-data -g crontab tee -a \
  /var/spool/cron/crontabs/www-data
echo "0  2  *  *  * /var/www/owncloud/occ dav:cleanup-chunks" \
  | sudo -u www-data -g crontab tee -a \
  /var/spool/cron/crontabs/www-data
----

[NOTE]
====

> If you need to sync your users from an LDAP or Active Directory Server, add this additional xref:configuration/server/background_jobs_configuration.adoc[Cron job]. Every 4 hours this cron job will sync LDAP users in ownCloud and disable the ones who are not available for ownCloud. Additionally, you get a log file in `/var/log/ldap-sync/user-sync.log` for debugging.

Reviewer 3: ```suggestion
If you need to sync your users from an LDAP or Active Directory Server, add this additional xref:configuration/server/background_jobs_configuration.adoc[cron job]. Every 4 hours this cron job will sync LDAP users in ownCloud and disable the ones who are not available for ownCloud. Additionally, you get a log file in `/var/log/ldap-sync/user-sync.log` for debugging.
```
